src/engine/IndexScan.cpp

```diff
@@ -335,44 +336,48 @@ IndexScan::getOptionalBlockSpan() const {
 
 // _____________________________________________________________________________
 std::optional<std::vector<CompressedBlockMetadata>>
-IndexScan::getOptionalPrefilteredBlocks(
-    std::vector<CompressedBlockMetadata> blocks) const {
-  if (!getLimit().isUnconstrained()) {
-    return std::nullopt;
+IndexScan::getBlockMetadataOptionallyPrefiltered() const {
+  auto optBlockSpan = getBlockMetadata();
+  std::optional<std::vector<CompressedBlockMetadata>> optBlocks = std::nullopt;
+  if (optBlockSpan.has_value()) {
+    const auto& blockSpan = optBlockSpan.value();
+    optBlocks = {blockSpan.begin(), blockSpan.end()};
+    applyPefilterIfPossible(optBlocks.value());
   }
+  return optBlocks;
+}
 
-  if (prefilter_.has_value()) {
+// _____________________________________________________________________________
+void IndexScan::applyPefilterIfPossible(
+    std::vector<CompressedBlockMetadata>& blocks) const {
+  if (prefilter_.has_value() && getLimit().isUnconstrained()) {
     // Apply the prefilter on given blocks.
     auto& [prefilterExpr, columnIndex] = prefilter_.value();
-    std::cout << "Use prefiltering" << std::endl;
-    return prefilterExpr->evaluate(blocks, columnIndex);
-  } else {
-    return blocks;
+    blocks = prefilterExpr->evaluate(blocks, columnIndex);
   }
 }
 
 // _____________________________________________________________________________
 Permutation::IdTableGenerator IndexScan::getLazyScan(
     std::vector<CompressedBlockMetadata> blocks) const {
+  applyPefilterIfPossible(blocks);
   // If there is a LIMIT or OFFSET clause that constrains the scan
   // (which can happen with an explicit subquery), we cannot use the prefiltered
   // blocks, as we currently have no mechanism to include limits and offsets
   // into the prefiltering (`std::nullopt` means `scan all blocks`).
-  auto actualBlocks = getOptionalPrefilteredBlocks(std::move(blocks));
-  return getIndex()
-      .getImpl()
-      .getPermutation(permutation())
-      .lazyScan(getScanSpecification(), std::move(actualBlocks),
-                additionalColumns(), cancellationHandle_,
-                locatedTriplesSnapshot(), getLimit());
+  auto filteredBlocks = getLimit().isUnconstrained()
+                            ? std::optional(std::move(blocks))
+                            : std::nullopt;
```

Comment:
Is that still necessary? I think that case is handled inside `applyPrefilterIfNecessary`?